### PR TITLE
Added data_extractor; new class to extract peak and hit information

### PR DESCRIPTION
**Pull request by ErikHogenbirk**
I made a branch to extract peak and hit information from the ROOT file, as opposed to the event level information that you can get with the normal minitrees. I can add a small bit of example code to examples if necessary.

### CODE REVIEW

File: hax/data_extractor.py [hunk 1]
```diff
@@ -0,0 +1,145 @@
+"""Extract peak or hit info from processed root file
+"""
+
+import numpy as np
+import hax
+
+def root_to_numpy(base_object, field_name, attributes):
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1

File: hax/data_extractor.py [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    """Convert objects stored in base_object.field_name to numpy array
+    Will query attributes for each of the objects in base_object.field_name
+    No, root_numpy does not do this for you, that's for trees...
+    """
+    objects_to_convert = getattr(base_object, field_name)
+    if not len(objects_to_convert):
+        return None
+    return np.array([tuple([getattr(p, pa) for pa in attributes]) for p in objects_to_convert])
+
+def build_cut_string(cut_list,obj):
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    '''
+    if len(cut_list) == 0:
+        return 'True'
+    cut_string = '('
+    for cut in cut_list[:-1]:
+        cut_string += obj + '.' + cut + ') & ('
+    cut_string += obj + '.' + cut_list[-1] + ')'
+    return cut_string
+
+def make_branch_selection(level, event_fields, peak_fields, added_branches):
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1

File: hax/data_extractor.py [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    """Make the list of branches that have to be selected.
+    """
+    branch_selection_events = event_fields
+    branch_selection_peaks = ['peaks.' + field for field in peak_fields]
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+    cut_string += obj + '.' + cut_list[-1] + ')'
+    return cut_string
+
+def make_branch_selection(level, event_fields, peak_fields, added_branches):
+    """Make the list of branches that have to be selected.
+    """
+    branch_selection_events = event_fields
+    branch_selection_peaks = ['peaks.' + field for field in peak_fields]
+    
+    # For hits, just select the whole hit branch.
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+def make_branch_selection(level, event_fields, peak_fields, added_branches):
+    """Make the list of branches that have to be selected.
+    """
+    branch_selection_events = event_fields
+    branch_selection_peaks = ['peaks.' + field for field in peak_fields]
+    
+    # For hits, just select the whole hit branch. 
+    # Unfortunately, specifically setting one variable does not seem to work.
+    branch_selection_hits = ['peaks.hits*']
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+    branch_selection_peaks = ['peaks.' + field for field in peak_fields]
+    
+    # For hits, just select the whole hit branch. 
+    # Unfortunately, specifically setting one variable does not seem to work.
+    branch_selection_hits = ['peaks.hits*'] 
+    if level == 'hit':
+        branch_selection = branch_selection_events + branch_selection_peaks + branch_selection_hits + added_branches
+    if level == 'peak':
+        branch_selection = branch_selection_events + branch_selection_peaks + added_branches
+    # Hack to enable range_50p_ar
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    
+    # For hits, just select the whole hit branch. 
+    # Unfortunately, specifically setting one variable does not seem to work.
+    branch_selection_hits = ['peaks.hits*'] 
+    if level == 'hit':
+        branch_selection = branch_selection_events + branch_selection_peaks + branch_selection_hits + added_branches
+    if level == 'peak':
+        branch_selection = branch_selection_events + branch_selection_peaks + added_branches
+    # Hack to enable range_50p_ar    
+    branch_selection = [field.replace('peaks.range_50p_area','peaks.range_area_decile*') for field in branch_selection]
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    return branch_selection
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    branch_selection_hits = ['peaks.hits*'] 
+    if level == 'hit':
+        branch_selection = branch_selection_events + branch_selection_peaks + branch_selection_hits + added_branches
+    if level == 'peak':
+        branch_selection = branch_selection_events + branch_selection_peaks + added_branches
+    # Hack to enable range_50p_ar    
+    branch_selection = [field.replace('peaks.range_50p_area','peaks.range_area_decile*') for field in branch_selection]
+    return branch_selection
+    
+def make_named_array(array,field_names):
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E302] expected 2 blank lines, found 1
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    """Make a named array from a numpy array.
+    """
+    import pandas as pd
+    df = pd.DataFrame(array, columns=field_names)
+    array = df.to_records()
+    return array
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    import pandas as pd
+    df = pd.DataFrame(array, columns=field_names)
+    array = df.to_records()
+    return array
+    
+
+class DataExtractor():
+    """This class is meant for extracting properties that are *not* on the event level, such as peak or hit properties.
+    """
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
...
+class DataExtractor():
+    """This class is meant for extracting properties that are *not* on the event level, such as peak or hit properties.
+    """
+    
+    def __init__(self):
+        # Initialize empty data list
+        self.data = []
+        # Default values for cuts: everything passes
+        event_cut_string = 'True'
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [F841] local variable 'event_cut_string' is assigned to but never used

File: hax/data_extractor.py [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+        peak_cut_string = 'True'
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [F841] local variable 'peak_cut_string' is assigned to but never used

File: hax/data_extractor.py [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+        # Default values for cuts: everything passes
+        event_cut_string = 'True'
+        peak_cut_string = 'True'
+        
+    def loop_body(self, event):
+        """Function that extracts data from each event and adds array with that data to the data list.
+        """
+        # Check if event passes event cut
+        if eval(self.event_cut_string):
+            event_entry = np.array([getattr(event,field) for field in self.event_fields])
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 17]
```diff
@@ -0,0 +1,145 @@
...
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        """Function that extracts data from each event and adds array with that data to the data list.
+        """
+        # Check if event passes event cut
+        if eval(self.event_cut_string):
+            event_entry = np.array([getattr(event,field) for field in self.event_fields])
+            
+            for peak in event.peaks:
+                # Check if peak passes the cut
+                if eval(self.peak_cut_string):
+                    # Get peak information
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+        """
+        # Check if event passes event cut
+        if eval(self.event_cut_string):
+            event_entry = np.array([getattr(event,field) for field in self.event_fields])
+            
+            for peak in event.peaks:
+                # Check if peak passes the cut
+                if eval(self.peak_cut_string):
+                    # Get peak information                
+                    #peak_entry = np.array([getattr(peak,field) for field in self.peak_fields])
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E265] block comment should start with '# '

File: hax/data_extractor.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+                    peak_entry = np.array([getattr(peak,field) if field is not 'range_50p_area' else
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+            event_entry = np.array([getattr(event,field) for field in self.event_fields])
+            
+            for peak in event.peaks:
+                # Check if peak passes the cut
+                if eval(self.peak_cut_string):
+                    # Get peak information                
+                    #peak_entry = np.array([getattr(peak,field) for field in self.peak_fields])
+                    peak_entry = np.array([getattr(peak,field) if field is not 'range_50p_area' else 
+                                           list(peak.range_area_decile)[5]
+                                           for field in self.peak_fields])
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+            
+            for peak in event.peaks:
+                # Check if peak passes the cut
+                if eval(self.peak_cut_string):
+                    # Get peak information                
+                    #peak_entry = np.array([getattr(peak,field) for field in self.peak_fields])
+                    peak_entry = np.array([getattr(peak,field) if field is not 'range_50p_area' else 
+                                           list(peak.range_area_decile)[5]
+                                           for field in self.peak_fields])                    
+                    if self.level == 'hit':
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+                # Check if peak passes the cut
+                if eval(self.peak_cut_string):
+                    # Get peak information                
+                    #peak_entry = np.array([getattr(peak,field) for field in self.peak_fields])
+                    peak_entry = np.array([getattr(peak,field) if field is not 'range_50p_area' else 
+                                           list(peak.range_area_decile)[5]
+                                           for field in self.peak_fields])                    
+                    if self.level == 'hit':                        
+                        # Extract hit info. Numpy array with n_hit_channel entries x number of hit properties
+                        hit_entry = root_to_numpy(peak,'hits',self.hit_fields)
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+                        if hit_entry is None :
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E203] whitespace before ':'

File: hax/data_extractor.py [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+                            # If there is no hit data: crash loudly
+                            raise ValueError("Unable to read hit info. Is it in the root file? (Try: peak_cuts=['type == 's1''])")
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E501] line too long (130 > 120 characters)

File: hax/data_extractor.py [hunk 26]
```diff
@@ -0,0 +1,145 @@
...
+
+                        # Append all properties. First event properties, then peak, then hits
+                        entry = np.c_[np.zeros((len(hit_entry),len(event_entry)+len(peak_entry))),hit_entry]
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 27]
```diff
@@ -0,0 +1,145 @@
...
+                        entry[:,0:len(event_entry)] = event_entry
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 28]
```diff
@@ -0,0 +1,145 @@
...
+                        entry[:,len(event_entry):(len(peak_entry)+len(event_entry))] = peak_entry
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 29]
```diff
@@ -0,0 +1,145 @@
...
+                    elif self.level == 'peak':
+                        # Extra brackets needed for concatenation in the end, else we'll get flat array
+                        entry = np.c_[[event_entry],[peak_entry]]
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 30]
```diff
@@ -0,0 +1,145 @@
...
+                    else:
+                        # We should actually never reach this since the checking has been done before
+                        raise SyntaxError("Enter either 'peak' of 'hit' for level!")
+                    self.data.append(entry)
+        # Check if user-defined event number limit is reached
+        if event.event_number >= self.stop_after:
+            print("User-defined limit of %d events reached, stopping..." % self.stop_after)
+            raise hax.paxroot.StopEventLoop
+        return None
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 31]
```diff
@@ -0,0 +1,145 @@
...
+                        # We should actually never reach this since the checking has been done before
+                        raise SyntaxError("Enter either 'peak' of 'hit' for level!")
+                    self.data.append(entry)
+        # Check if user-defined event number limit is reached
+        if event.event_number >= self.stop_after:
+            print("User-defined limit of %d events reached, stopping..." % self.stop_after)
+            raise hax.paxroot.StopEventLoop
+        return None
+        
+    def get_data(self, dataset, level ='peak', event_fields = ['event_number'], peak_fields=['area','hit_time_std'], hit_fields = [],
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E251] unexpected spaces around keyword / parameter equals
* [E231] missing whitespace after ','
* [E501] line too long (133 > 120 characters)

File: hax/data_extractor.py [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+                 event_cuts=[], peak_cuts=[], stop_after = np.inf, added_branches=[]):
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E251] unexpected spaces around keyword / parameter equals

File: hax/data_extractor.py [hunk 33]
```diff
@@ -0,0 +1,145 @@
...
+        """Extract peak or hit data from a dataset.
+        Peak or hit can be toggled by specifying level = 'peak' or level = 'hit'.
+        Example useage:
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 34]
```diff
@@ -0,0 +1,145 @@
...
+            print("User-defined limit of %d events reached, stopping..." % self.stop_after)
+            raise hax.paxroot.StopEventLoop
+        return None
+        
+    def get_data(self, dataset, level ='peak', event_fields = ['event_number'], peak_fields=['area','hit_time_std'], hit_fields = [],
+                 event_cuts=[], peak_cuts=[], stop_after = np.inf, added_branches=[]):
+        """Extract peak or hit data from a dataset.
+        Peak or hit can be toggled by specifying level = 'peak' or level = 'hit'.
+        Example useage: 
+            d = de.get_data(dataset=run_name,level='peak',event_fields = ['event_number'],
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 35]
```diff
@@ -0,0 +1,145 @@
...
+            d = de.get_data(dataset=run_name,level='peak',event_fields = ['event_number'], 
+                peak_fields=['area'],event_cuts=['event_number > 5', 'event_number < 10'],
+                peak_cuts=['area > 100', 'type = "s1"'],stop_after=10000,added_branches= ['peak.type'])
+        """
+        # Sanity checking
+        if (level != 'peak') and (level != 'hit'):
+            raise SyntaxError("Enter either 'peak' of 'hit' for level!")
+        if (hit_fields != []) and (level == 'peak'):
+            print("Warning: You set hit properties, but your input will be ignored since you specified peak level!")
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 36]
```diff
@@ -0,0 +1,145 @@
...
+                peak_fields=['area'],event_cuts=['event_number > 5', 'event_number < 10'],
+                peak_cuts=['area > 100', 'type = "s1"'],stop_after=10000,added_branches= ['peak.type'])
+        """
+        # Sanity checking
+        if (level != 'peak') and (level != 'hit'):
+            raise SyntaxError("Enter either 'peak' of 'hit' for level!")
+        if (hit_fields != []) and (level == 'peak'):
+            print("Warning: You set hit properties, but your input will be ignored since you specified peak level!")
+        
+        branch_selection = make_branch_selection(level,event_fields,peak_fields,added_branches)
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

**Comment by ErikHogenbirk** [comment]
Wow. There are a *lot* of style errors. As they say, 'it's my first day', will fix :)

File: hax/data_extractor.py [hunk 37]
```diff
@@ -0,0 +1,145 @@
...
+        self.event_cut_string = build_cut_string(event_cuts,'event')
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 38]
```diff
@@ -0,0 +1,145 @@
...
+        self.peak_cut_string = build_cut_string(peak_cuts,'peak')
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 39]
```diff
@@ -0,0 +1,145 @@
...
+        self.event_fields = event_fields
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 40]
```diff
@@ -0,0 +1,145 @@
...
+        if (level != 'peak') and (level != 'hit'):
+            raise SyntaxError("Enter either 'peak' of 'hit' for level!")
+        if (hit_fields != []) and (level == 'peak'):
+            print("Warning: You set hit properties, but your input will be ignored since you specified peak level!")
+        
+        branch_selection = make_branch_selection(level,event_fields,peak_fields,added_branches)
+        self.event_cut_string = build_cut_string(event_cuts,'event')
+        self.peak_cut_string = build_cut_string(peak_cuts,'peak')
+        self.event_fields = event_fields        
+        self.peak_fields = peak_fields
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 41]
```diff
@@ -0,0 +1,145 @@
...
+            raise SyntaxError("Enter either 'peak' of 'hit' for level!")
+        if (hit_fields != []) and (level == 'peak'):
+            print("Warning: You set hit properties, but your input will be ignored since you specified peak level!")
+        
+        branch_selection = make_branch_selection(level,event_fields,peak_fields,added_branches)
+        self.event_cut_string = build_cut_string(event_cuts,'event')
+        self.peak_cut_string = build_cut_string(peak_cuts,'peak')
+        self.event_fields = event_fields        
+        self.peak_fields = peak_fields 
+        #self.peak_fields = [field.replace('range_50p_area','range_area_decile[5]') for field in peak_fields]
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E265] block comment should start with '# '

File: hax/data_extractor.py [hunk 42]
```diff
@@ -0,0 +1,145 @@
...
+        self.hit_fields = hit_fields
+        self.stop_after = stop_after
+        self.level = level
+
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W293] blank line contains whitespace

File: hax/data_extractor.py [hunk 43]
```diff
@@ -0,0 +1,145 @@
...
+        self.event_cut_string = build_cut_string(event_cuts,'event')
+        self.peak_cut_string = build_cut_string(peak_cuts,'peak')
+        self.event_fields = event_fields        
+        self.peak_fields = peak_fields 
+        #self.peak_fields = [field.replace('range_50p_area','range_area_decile[5]') for field in peak_fields]
+        self.hit_fields = hit_fields
+        self.stop_after = stop_after
+        self.level = level
+        
+        hax.paxroot.loop_over_dataset(dataset,self.loop_body,branch_selection=branch_selection)
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 44]
```diff
@@ -0,0 +1,145 @@
...
+
+        # Now reshape data
+        # list of arrays -> one array -> named array
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W291] trailing whitespace

File: hax/data_extractor.py [hunk 45]
```diff
@@ -0,0 +1,145 @@
...
+        hax.paxroot.loop_over_dataset(dataset,self.loop_body,branch_selection=branch_selection)
+
+        # Now reshape data
+        # list of arrays -> one array -> named array           
+        self.data = np.concatenate(self.data)
+        # Build list of strings with field names.
+        if level == 'hit':
+            # For the hit level, we have to be careful. For example, 'area' can be peak or hit area.
+            # How to solve this? Well, just add hit_ or peak_ before the property
+            field_names = event_fields + ['peak_' + field for field in peak_fields] + ['hit_' + field for field in hit_fields]
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E501] line too long (126 > 120 characters)

File: hax/data_extractor.py [hunk 46]
```diff
@@ -0,0 +1,145 @@
...
+        if level == 'peak':
+            # For peak level no such problem exists (yet) so just keep normal names
+            field_names = event_fields + peak_fields
+        self.data = make_named_array(self.data,field_names)
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [E231] missing whitespace after ','

File: hax/data_extractor.py [hunk 47]
```diff
@@ -0,0 +1,145 @@
...
+
+        return self.data
```
Comment (tunnell) [inline hax/data_extractor.py]: __[flake8]__
* [W292] no newline at end of file

**Comment by JelleAalbers** [comment]
We don't care about the style, hax is not pep8 compliant. At least I don't think so... Chris just left this webhook thing on, sorry to everyone who got emails. I've turned it off now.

**Comment by ErikHogenbirk** [comment]
Thanks. I changed most of them now anyway.

**Comment by JelleAalbers** [comment]
Thanks a lot Erik! I think we can improve the interface a bit, but since it is such an important feature, it works well, and I'm already using it :-) I'm going to merge. Chris can slap me when he gets back. 

People will just have to expect the API for the feature to change. Right now it's undocumented anyway.

**Pull request merged**